From b2e41623b8fb7bb10fd257cb7b2f52d3a8037986 Mon Sep 17 00:00:00 2001
From: Pavel Zakharov <pavel.zakharov@delphix.com>
Date: Tue, 19 Jun 2018 13:14:27 -0400
Subject: [PATCH] 
 0027-Set-default-sshd-options-based-on-etc-default-login.patch

---
 pathnames.h   |  1 +
 servconf.c    | 61 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 sshd_config.4 | 17 +++++++++++++++--
 3 files changed, 77 insertions(+), 2 deletions(-)

diff --git a/pathnames.h b/pathnames.h
index 4fa9a10..feef56b 100644
--- a/pathnames.h
+++ b/pathnames.h
@@ -47,6 +47,7 @@
 #define _PATH_HOST_XMSS_KEY_FILE	SSHKEYDIR "/ssh_host_xmss_key"
 #define _PATH_HOST_RSA_KEY_FILE		SSHKEYDIR "/ssh_host_rsa_key"
 #define _PATH_DH_MODULI			SSHKEYDIR "/moduli"
+#define _PATH_DEFAULT_LOGIN		ETCDIR "/default/login"
 /* Backwards compatibility */
 #define _PATH_DH_PRIMES			SSHDIR "/primes"
 
diff --git a/servconf.c b/servconf.c
index c8d125c..75de759 100644
--- a/servconf.c
+++ b/servconf.c
@@ -36,6 +36,7 @@
 #include <unistd.h>
 #include <limits.h>
 #include <stdarg.h>
+#include <deflt.h>
 #include <errno.h>
 #ifdef HAVE_UTIL_H
 #include <util.h>
@@ -199,6 +200,64 @@ option_clear_or_none(const char *o)
 	return o == NULL || strcasecmp(o, "none") == 0;
 }
 
+/*
+ * Reads /etc/default/login and defaults several ServerOptions:
+ *
+ * PermitRootLogin
+ * PermitEmptyPasswords
+ * LoginGraceTime
+ *
+ * CONSOLE=*      -> PermitRootLogin=without-password
+ * #CONSOLE=*     -> PermitRootLogin=yes
+ *
+ * PASSREQ=YES    -> PermitEmptyPasswords=no
+ * PASSREQ=NO     -> PermitEmptyPasswords=yes
+ * #PASSREQ=*     -> PermitEmptyPasswords=no
+ *
+ * TIMEOUT=<secs> -> LoginGraceTime=<secs>
+ * #TIMEOUT=<secs> -> LoginGraceTime=300
+ */
+static void
+deflt_fill_default_server_options(ServerOptions *options)
+{
+	int	flags;
+	char	*ptr;
+
+	if (defopen(_PATH_DEFAULT_LOGIN))
+		return;
+
+	/* Ignore case */
+	flags = defcntl(DC_GETFLAGS, 0);
+	TURNOFF(flags, DC_CASE);
+	(void) defcntl(DC_SETFLAGS, flags);
+
+	if (options->permit_root_login == PERMIT_NOT_SET &&
+	    (ptr = defread("CONSOLE=")) != NULL)
+		options->permit_root_login = PERMIT_NO_PASSWD;
+
+	if (options->permit_empty_passwd == -1 &&
+	    (ptr = defread("PASSREQ=")) != NULL) {
+		if (strcasecmp("YES", ptr) == 0)
+			options->permit_empty_passwd = 0;
+		else if (strcasecmp("NO", ptr) == 0)
+			options->permit_empty_passwd = 1;
+	}
+
+	if (options->max_authtries == -1 &&
+	    (ptr = defread("RETRIES=")) != NULL) {
+		options->max_authtries = atoi(ptr);
+	}
+
+	if (options->login_grace_time == -1) {
+		if ((ptr = defread("TIMEOUT=")) != NULL)
+			options->login_grace_time = (unsigned)atoi(ptr);
+		else
+			options->login_grace_time = 300;
+	}
+
+	(void) defopen((char *)NULL);
+}
+
 static void
 assemble_algorithms(ServerOptions *o)
 {
@@ -262,6 +321,8 @@ fill_default_server_options(ServerOptions *options)
 		options->use_pam = 0;
 #endif
 
+	deflt_fill_default_server_options(options);
+
 	/* Standard Options */
 	if (options->num_host_key_files == 0) {
 		/* fill default hostkeys for protocols */
diff --git a/sshd_config.4 b/sshd_config.4
index 441f043..fb2df91 100644
--- a/sshd_config.4
+++ b/sshd_config.4
@@ -1155,7 +1155,13 @@ Specifies the maximum number of authentication attempts permitted per
 connection.
 Once the number of failures reaches half this value,
 additional failures are logged.
-The default is 6.
+The default is 6, or the value given by
+.Dq RETRIES=
+in the file
+.Dq /etc/default/login ,
+if available (see
+.Xr login 1
+).
 .It Cm MaxSessions
 Specifies the maximum number of open shell, login or subsystem (e.g. sftp)
 sessions permitted per network connection.
@@ -1206,7 +1212,14 @@ The default is
 When password authentication is allowed, it specifies whether the
 server allows login to accounts with empty password strings.
 The default is
-.Cm no .
+.Cm no
+unless
+.Dq PASSREQ=YES
+is present in
+.Dq /etc/default/login
+(see
+.Xr login 1
+).
 .It Cm PermitOpen
 Specifies the destinations to which TCP port forwarding is permitted.
 The forwarding specification must be one of the following forms:
-- 
2.14.3 (Apple Git-98)

